src/server/scene/basic_surface.h

```diff
@@ -156,7 +156,15 @@ class BasicSurface : public Surface
     void set_application_id(std::string const& application_id) override;
 
 private:
-    bool visible(std::lock_guard<std::mutex> const&) const;
+    struct ProofOfMutexLock
+    {
+        ProofOfMutexLock(std::lock_guard<std::mutex> const&) {}
+        ProofOfMutexLock(std::unique_lock<std::mutex> const&) {}
```

Comment:
Note that this doesn't actually prove the lock is held; 'std::unique_lock`s can be unlocked.

Reply:
The way to fix this would be something along the lines of `if (!lock) { lock.lock() }`

Reply:
`if (!lock) { lock.lock(); }` would not be safe because on the rare occasion the current thread did not hold the lock but another thread did, it would assume the lock was held by the current thread and not obtain it. Perhaps we could throw in the `unique_lock` constructor if the lock is not held.

Reply:
The only way that could happen is if the `std::unique_lock` instance was shared between threads, which is (a) possible, but (b) very much the wrong thing to do, and is *also* possible with `std::lock_guard`.

That said, probably what you *actually* want is a `MutexGuard<State>`. That's a rather more expansive change, though.

Reply:
Ah, I see what you mean. You *could* safely lock the lock, but I think trying to create a proof of a lock without already having a lock should be a `std::logic_error`.